src/main/kotlin/org/opensearch/replication/task/index/IndexReplicationTask.kt

```diff
@@ -163,76 +164,91 @@ open class IndexReplicationTask(id: Long, type: String, action: String, descript
+                        }
+
+                        if (state !is MonitoringState) {
+                            // Tasks need to be started
+                            state
+                        } else {
+                            state = pollShardTaskStatus((followingTaskState as FollowingState).shardReplicationTasks)
+                            followingTaskState = startMissingShardTasks((followingTaskState as FollowingState).shardReplicationTasks)
+                            when (state) {
+                                is MonitoringState -> {
+                                    updateMetadata()
+                                }
+                                is FailedState -> {
+                                    // Try pausing first if we get Failed state. This returns failed state if pause failed
+                                    pauseReplication(state)
+                                }
+                                else -> {
+                                    state
+                                }
                             }
                         }
                     }
+                    ReplicationState.FAILED -> {
+                        assert(currentTaskState is FailedState)
+                        failReplication(currentTaskState as FailedState)
+                        currentTaskState
+                    }
+                    ReplicationState.COMPLETED -> {
+                        markAsCompleted()
+                        CompletedState
+                    }
                 }
-                ReplicationState.FAILED -> {
-                    assert(currentTaskState is FailedState)
-                    failReplication(currentTaskState as FailedState)
-                    currentTaskState
-                }
-                ReplicationState.COMPLETED -> {
-                    markAsCompleted()
-                    CompletedState
+                if (newState != currentTaskState) {
+                    currentTaskState = updateState(newState)
                 }
+                if (isCompleted) break
             }
-            if (newState != currentTaskState) {
-                currentTaskState = updateState(newState)
+            catch(e: OpenSearchException) {
+                val status = e.status().status
+                // Index replication task shouldn't exit before shard replication tasks
+                // As long as shard replication tasks doesn't encounter any errors, Index task
+                // should continue to poll and Any failure encoutered from shard task should
+                // invoke state transition and exit
+                if(status < 500 && status != RestStatus.TOO_MANY_REQUESTS.status) {
+                    log.error("Exiting index replication task", e)
+                    throw e
+                }
+                log.debug("Encountered transient error while running index replication task", e)
+                delay(SLEEP_TIME_BETWEEN_POLL_MS)
```

Comment:
Again, exponential backoff? 
Also, should we limit the number of retries here?

Reply:
Thought of this but it will only space out the calls for that duration any intermediate shard task updates might be missed during that prolonged wait and 
for this case, we don't want to exit for 5xx errors even after the retries. 

so, kept it simple. Let me know, If we you think otherwise.